### ISSUE TITLE
RFC Fusing together application and record manipulation syntaxes

### ISSUE THREAD

**Issue by Russoul** [idris-lang/Idris2#642]
This is another take on #626 

We look at the problem from a different perspective, trying to solve a few known connected problems at once
in a uniform style.

### SYNTAX PROPOSAL

We have 3 groups of arguments: explicit, implicit, auto implicit.
Each of them can be named or unnamed.

### RULES

for both LHS and RHS:

- Positions of named arguments are irrelevant.
- In case of duplicate named arguments, disambiguate by using group exclusive syntax (defined below).
- Unnamed arguments are always positional relative to all other from the same group
- Positions of unnamed arguments relative to all other arguments from another groups are irrelevant.
- On LHS unnamed patterns bind named arguments but named patterns are prioritized
- On RHS unnamed arguments get passed to named positions, but named arguments are prioritized

### APPLICATION SYNTAX

New syntax allows grouping multiple arguments in one set of braces.

Named variants:

- `{x = ..., y = ..., ...}` for all named arguments
- `${x = ..., y = ..., ...}` for named implicit only, handy in case of duplicate names in Pi binders
- `@{x = ..., y = ..., ...}` for named auto implicit only, same logic

`{x, y, ...} = {x = x, y = y, ...}`  pun any named, same for implicit and auto implicit variants

Unnamed variants:
`...` for explicit
`${..., ...}` for implicit
`@{..., ...}` for auto implicit

### NEW HOLE VARIANT

Please do not confuse `...` (external syntax) with `..` (internal, proposed syntax) below:

`..` - a hole standing for a data constructor such that: 

1. it (constructor) is the only one for an associated data type
2. _or_ other adjacent constructors are rejected by unification

Another proposed syntax for the hole: `_!` .

`..` is useful on its own, to reduce noise from names in some places.
But it has another important use: one can reference machine generated constructors of any data/record/interface type
if they follow the terms of use of the `..`

We exploit this feature in the following way:

### NEW DATA/RECORD/INTERFACE UPDATE SYNTAX:

We remove the redundant and misleading 'record' keyword,
which results in uniform, cleaner syntax (notice `:=` instead of `=`):

- `{x := x'}` is interpreted as: `\(.. bindEverythingButX x) => .. bindEverythingButX x'`
- `{x $= f}` is interpreted as: `\(.. bindEverythingButX x) => .. bindEverythingButX (f x))`

### PUTTING IT ALL TOGETHER
All definitions below are intended to pass typechecking if not said otherwise:

```idris
func0 : (x : Int) -> {x : Int} -> ()

-- in such cases: each named argument in application
-- must uniquely and independently of others identify its position

call00 : ()
call00 = func0 {x = 2} 1 -- won't typecheck, position of {x = 2} depends on position of '1'

call01 : ()
call01 = func0 ${2} 1

call02 : ()
call02 = func0 ${x = 2} 1

call03 : ()
call03 = func0 ${x = 2} {x = 1} -- won't typecheck, position of {x = 1} depends on position of ${x = 2}

call04 : ()
call04 = func0 {x = 2} {x = 1} -- won't typecheck, both arguments do not uniquely identify their position

call05 : ()
call05 = func0 {x = 2} -- won't typecheck, {x = 2} does not uniquely identify its position

func1 : Int -> (x : Int) -> {_ : Int} -> {y : Int} -> {auto _ : Int} -> {auto z : Int} -> ()

 -- all arguments keep their positions in the next three definitions:

call10 : ()
call10 = func1 1 {x = 2} ${3} {y = 4} @{5} {z = 6}

call11 : ()
call11 = func1 1 2 ${3} {y = 4} @{5} {z = 6}

call12 : ()
call12 = func1 {x = 2, y = 4, z = 6} @{5} ${3} 1


record Record a where
   constructor MkRecord
   {auto x : a}
   {y : a}
   z : a

interface (conA : ConstraintA a) => ConstraintB b => Interface a b where
   map : a -> b

makeRecord : Record Int
makeRecord = MkRecord {z = 3, x = 1, y = 2} -- now the 'record syntax' works naturally
                                            -- and without the redundant 'record' keyword

makeRecord' : Record Int
makeRecord' = .. {z = 3, x = 1, y = 2}     -- we are able to omit the constructor here

getZ : Record a -> a
getZ (.. {z}) = z

getZ' : Record a -> a
getZ' (.. z) = z

makeInterface : (conA : ConstraintA Int) => ConstraintB Int => Interface Int Int
makeInterface @{_} @{conB} = .. {conA, conB, map = ...} -- we still have to match on 'conA'
                                                        -- in order to access the second unnamed auto implicit
                                                        -- this hasn't changed from current Idris behaviour

implies : (x : Bool) -> (y : Bool) -> Bool
implies {y = False} {x = True} = False
implies {y = _} _ {- x -} = True -- 'y' pattern takes priority over the second explicit match-any pattern
                                 -- regardless of position of 'x' and 'y' in the type signature
                                 -- after 'y' gets bound, the match-any binds the first unbound argument, i.e. 'x'

priority : (a : Int) -> (b : Int) -> (c : Int) -> Int -> ()
priority {c} a {b} d = ()        -- same logic

manyArgs : (onMouseButton : SameType) -> (onKeyPressed : SameType) -> (onMouseMotion : SameType) -> ...


niceLessBugs : ...
niceLessBugs = manyArgs { onKeyPressed  = ... -- cleanly provide explicit
                        , onMouseMotion = ... -- arguments by name
                        , onMouseButton = ...
                        , ... }


func2 : Int -> {_ : Int} -> {x : Int} -> {auto _ : Int} -> {auto x : Int} -> ()

call2 : ()
call2 = @{x = 5} @{4} ${x = 3} ${2} 1

data ManyArgs : Type where
     MkManyArgs : Int -> {_ : Int} -> {x : Int} -> {auto _ : Int} -> {auto x : Int} -> ManyArgs

getFourth : ManyArgs -> Int
getFourth (.. _ @{fourth}) = fourth -- we still have to bind all explicits on LHS

interface Vec2 a where
   constructor MkVec2
   x : a
   y : a

map : (a -> b) -> Vec2 a -> Vec2 b
map f = {x $= f, y $= f} -- clean, the 'record' keyword doesn't confuse anymore

setX : a -> Vec2 a -> Vec2 a
setX x' = {x := x'}

data EmptyOrNot : (xs : List a) -> Type where
   Empty : EmptyOrNot (x :: xs)
   NotEmpty : EmptyOrNot []

emptiness : (xs : List a) -> EmptyOrNot xs
emptiness (x :: xs) = .. -- only one constructor is applicable here
emptiness []        = .. -- same

data SBool : Bool -> Type where
     SFalse : SBool False
     STrue  : SBool True

sNot : (Subset Bool SBool) -> ... -- partially blocked by #499 (matching on irrelevant erased constructors)
sNot (.. False ..) = ... -- 'Element' is inferred, so is the erased 'SFalse' value (and we can use this information on RHS)
sNot (.. True ..)  = ... -- 'Element' is inferred, so is the erased 'STrue' value (and we can use this information on RHS)
                         -- less noise, all redundant information is inferred

interface Vec4 a where
   constructor MkVec4
   x : a
   y : a
   z : a
   w : a

useVec4 : Monad m => Vec4 a -> m b
useVec4 (.. x y z w) = ...   -- again, we have to match on all explicit arguments of Vec4, even if we don't use all of them on RHS
                             -- but it might be beneficial:
                             -- 1) if one knows that they won't use most of the arguments of a data/record/interface type
                             --    they'd better use field projections
                             -- 2) given Idris 2 cultivates the usage of linearity
                             --    fully breaking down a data type which is linearly bound is helpful,
                             --    as we are required to use each argument exactly once



mapVect : (a -> b) -> Vect k a -> Vect k b
mapVect _ [] = ..
mapVect f v@(_ :: _) = {x $= f, xs $= mapVect f} v -- because why not
                                                   -- idris knows here that (::) is the only possible constructor
                                                   -- thus '..' can legitimately be used here
                                                   -- thus update syntax is accepted

-- mapVect f v@(_ :: _) = (\(x :: xs) => f x :: mapVect f xs) v -- this is what we get after elaborating {... $= ...}
```

### IN SUMMARY

- We reuse existing familiar syntax.
- We fuse together syntaxes for applications and record manipulations.
- We add a way to:
       1.       supply functions with unnamed implicit arguments
       2.       provide explicit arguments by name
       3.       provide multiple named arguments at once

- It seems we don't break much, except maybe for the record updates.
   We can solve this by leaving the old update syntax intact, deprecating it and adding the new one.

The purpose of this proposal is to unify Idris syntax for applications and records.
Making applications more user friendly and naturally supporting "record syntax".

### EXTRA SYNTAX IDEA (can be implemented later or disregarded fully)

`(_)` plays a role similar to `_` in bind mode:

     When the unifier doesn't have information about what an `_` hole should be instantiated with,
       it creates an implicit Pi binder and replaces that hole with a variable referencing it.

`(_)` would work the same way, but the typechecker would always create, in this a case, a lambda binder and replace `(_)` with a referencing variable.

Each occurrence of `(_)` is unique and is used exactly once and in order it was introduced (same as `_` in bind mode)
Thus: `(_) + (_)` elaborates to `\x, y => x + y` .
`(_)` allows one to shorten simple lambda expressions and avoid introducing extra names into scope.

```idris
mapMaybe : Maybe a -> (a -> b) -> Maybe b
mapMaybe Nothing = const Nothing
mapMaybe (Just x) = Just ((_) x)

useLinearOrdered : (1 _ : (1 _ : a) -> (1 _ : b) -> c) -> (1 _ : a) -> (1 _ : b) -> c
useLinearOrdered = (_) (_) (_)
-- useLinearOrdered = \f, x, y => f x y
```

Out of context use cases:
```idris
filter (\x => not (f x `elem` xs))
```
becomes
```idris
filter (not (f (_) `elem` xs))
```
And
```idris
newNameAndAge (\name', age' => Animal name' age' field3 field4 field5 field6) animal
```
becomes
```idris
newNameAndAge (Animal (_) (_) field3 field4 field5 field6) animal
```

Such patterns are common and misplaced or misspelled names in lambda expressions is the source of bugs.
Having `(_)` is a nice and simple solution.

**Comment by Russoul** [comment]
@gallais It would be very helpful to hear your opinion on this RFC and #626, we need to choose which way to go :)

**Comment by Russoul** [comment]
I've recently checked out this PR #314. I believe such projects can greatly benefit from the proposed syntax.
When working with generic constructions we often debate over the best approach to explicitness (or implicitness) of `Pi`s. Implicit arguments are great in proofs. It's nice when the compiler can infer information for us and we don't have to explicitly insert a bunch of `_` just to tell it our intention. But sometimes we have to nudge it a bit towards the right types. But the more information we provide, the noisier it gets. The new syntax can decrease the noise from braces when consecutive implicit arguments are provided:
```
... someProofFunc {a = ..., b = ..., c = ...}
```
vs
```
... someProofFunc {a = ...} {b = ...} {c = ...}
```

Maybe that would end the struggle of explicit vs implicit.

**Comment by ziman** [comment]
I really like the `{a = ..., b = ...}` syntax for any named argument(s) (implicit or explicit) of any function. Record patterns (`getBar (Foo {bar, baz}) = bar`) look okay to me, too, although the outer brackets feel a bit redundant/noise. Also, extending `@{...}` with support for multiple fields would be a nice side effect, but probably not used frequently.

On the other hand:
* I can't see a reason to have `${...}` when `{...}` already does its job. Yes, there is *some* disambiguation but you can already have two arguments of the same plicity that are named the same, so if you want to solve that problem, you'd have to come up with something more powerful, anyway.
* `..` looks strange to me; we already have `_` for "don't care", `..` evokes a range to me, and `_` can't appear in the pattern application head position now, anyway, can it? So the syntax cannot clash.
* we'd still have to use *data* constructors for records, as opposed to their type constructors. That's probably not too horrible; i think having a universal way to name arguments of *any* function outweighs this, and we can still get some extra sugar specific to records later, if we choose to, anyway.

So in short, we can get pretty much all of the value (that matters to me) just by implementing the `{...}` syntax in applications and patterns, and nothing else. Of course, this is a subjective statement.

**Comment by Russoul** [comment]
> Record patterns (getBar (Foo {bar, baz}) = bar) look okay to me, too, although the outer brackets feel a bit redundant/noise

@ziman You know what's nice ? We can add `.{` as you proposed. We get double benefit having it working with applications in general.
`getBar (Foo {bar, baz})` -> `getBar Foo.{bar, baz}` (head and args not reversed)
`record {age $= (+1), name = "Nick"} person` -> `{age $= (+1), name := "Nick"} person` -> `person.{age $= (+1), name := "Nick"}` (head and arg reversed)

> I can't see a reason to have ${...} when {...} already does its job

Let's trash `${...}` for now !

> `..` looks strange to me

Best I could find in terms of symbols more or less resembling `_`

Let's postpone `..`. It can become an independent project on its own. `..` can be used in argument position, that's a clash.
Same for `(_)`.

So how about we take one step and implement: general `{x1 = t1, x2 = t2, ...}`, `.{x1 = t1, x2 = t2, ...}`.
And  `{x1 $= t1, x2 := t2, ...}`, `.{x1 $= t1, x2 := t2, ...}`  which (the later two) are purely syntactic, aren't they ?
Now that we are about to reserve `:=` anyway with #691. That deprecates uses of bloated `record` keyword in terms. Should be an effortless addition (hopefully).

@ziman I love that all your original ideas from #626 smoothly incorporate into this. Except for implied constructors, which correlate with `..`. That would be step two :)

Basically that is an enhanced mix of what @ziman and I have come up with :)

@gallais @ziman @ohad What do you think ?

**Comment by ohad** [comment]
I don't have any additional points about the current proposal --- it looks amazing!

I'm not quite sure what 'We exploit this feature in the following way:' refers to, but I like having this kind of lightweight substitution operation.

Re. `..`: it seems weird to trash (or keep) it only based on surface syntax. If we don't like the syntax, but we like the idea, we should just choose something else. 
`_.` ("it can only be one thing, which you can work out from the other dependencies"), or 
`_!` ("work it out, it's unique") which is already in the proposal.
`|_` ("there is only one case", but might cause ambiguity)
or anything else that's acceptable.

**Comment by Russoul** [comment]
> I don't have any additional points about the current proposal --- it looks amazing!

Glad to hear !

> I'm not quite sure what 'We exploit this feature in the following way:' refers to

I was talking about update syntax utilising `..` under the hood:

  - {x := x'} is interpreted as: \(`..` bindEverythingButX x) => `..` bindEverythingButX x'
  - {x $= f} is interpreted as: \(`..` bindEverythingButX x) => `..` bindEverythingButX (f x))

That makes updates even more powerful, allowing stuff like:
```idris
mapVect : (a -> b) -> Vect k a -> Vect k b
mapVect _ [] = ..
mapVect f v@(_ :: _) = {x $= f, xs $= mapVect f} v
```
Where we weaken restrictions of forms of an update subject, in this case it is `Vect`. `..` is uniquely inferred to `Nil` and `(::)` in first and second branches accordingly (second branch utilising it implicitly inside update syntax)

**Comment by gallais** [comment]
I guess this can be closed. We have #965 to track the ripple effects.
